hotshot-query-service/src/availability.rs

```diff
@@ -161,9 +166,97 @@ impl Error {
     }
 }
 
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(bound = "")]
+pub struct Leaf1QueryData<Types: NodeType> {
+    pub(crate) leaf: Leaf<Types>,
+    pub(crate) qc: QuorumCertificate<Types>,
+}
+
+fn downgrade_leaf<Types: NodeType>(leaf2: Leaf2<Types>) -> Leaf<Types> {
+    // TODO do we still need some check here?
+    // `drb_seed` no longer exists on `Leaf2`
+    // if leaf2.drb_seed != [0; 32] && leaf2.drb_result != [0; 32] {
+    //     panic!("Downgrade of Leaf2 to Leaf will lose DRB information!");
+    // }
+    let quorum_proposal = QuorumProposal {
+        block_header: leaf2.block_header().clone(),
+        view_number: leaf2.view_number(),
+        justify_qc: leaf2.justify_qc().to_qc(),
+        upgrade_certificate: leaf2.upgrade_certificate(),
+        proposal_certificate: None,
+    };
+    let mut leaf = Leaf::from_quorum_proposal(&quorum_proposal);
+    if let Some(payload) = leaf2.block_payload() {
+        leaf.fill_block_payload_unchecked(payload);
+    }
+    leaf
+}
+
+impl<Types: NodeType> From<LeafQueryData<Types>> for Leaf1QueryData<Types> {
+    fn from(value: LeafQueryData<Types>) -> Self {
```

Comment:
I think we should **not** implement `From` when the conversion is lossy. This is the second bullet point here: https://doc.rust-lang.org/std/convert/trait.From.html#when-to-implement-from

I think if the conversion is lossy it's better to have a function or method, this then also makes it easier to search the codebase for these lossy conversions.

cc @imabdulbasit

Reply:
https://github.com/EspressoSystems/espresso-sequencer/pull/2478/commits/0d272a30b9db0d86399188d83db4c0624b88a070